runtime/src/main/java/org/corfudb/runtime/collections/StreamPollingTask.java

```diff
@@ -0,0 +1,108 @@
+import org.corfudb.runtime.view.stream.IStreamView;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A runnable task for a subscription to poll data change from the
+ * transaction stream and put into the subscription's buffer. This
+ * task is executed by the thread pool continuously until an error
+ * occurs which would stop the subscription.
+ * <p>
+ * Created by WenbinZhu on 11/9/20.
+ */
+@Slf4j
+class StreamPollingTask implements Runnable {
+
+    // A period of time in ms to sleep before next cycle when poller gets no new data changes.
+    private static final Duration IDLE_WAIT_TIME_MS = Duration.ofMillis(50);
+
+    // Total amount of time to wait for putting the polled data changes into buffer if it is full.
+    private static final Duration QUEUE_FULL_BLOCK_TIME_MS = Duration.ofMillis(2_000);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The corfu transaction stream to poll data changes from.
+    private final IStreamView txnStream;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing stream polling tasks.
+    private final ExecutorService pollingExecutor;
+
+    // Last address of the data successfully processed by the buffer.
+    private long lastReadAddress;
+
+    StreamPollingTask(StreamingManager streamingManager, long lastAddress,
+                      StreamSubscription subscription, ExecutorService executor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.pollingExecutor = executor;
+        this.lastReadAddress = lastAddress;
+        this.txnStream = subscription.getTxnStream();
+    }
+
+    @Override
+    public void run() {
+        try {
+            pollTxnStream();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during txn stream polling, listener: {}, " +
+                    "namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Poll new data changes from the transaction stream and put into
+     * the subscription's buffer one by one.
+     */
+    private void pollTxnStream() throws Exception {
+        // If listener already unsubscribed, do not process or schedule again.
+        if (subscription.isStopped()) {
+            return;
+        }
+
+        // Seek to next address and poll transaction updates.
+        txnStream.seek(lastReadAddress + 1L);
+        List<ILogData> updates = txnStream.remainingAtMost(subscription.getStreamBufferSize());
+
```

Comment:
The logic can be simplified while also improving graceful degradation by actually trying to check if `StreamSubscription::streamBuffer` has enough space before issuing the read to the server. Otherwise if the system is in a degraded state and the buffers are not being consumed fast enough, then you will be issuing reads and computing all the logic in `StreamSubscription::enqueueStreamEntry`  and then returning false on `streamBuffer.offer`. 

When throttling is actually required its probably better assume the pessimistic case vs the optimistic case (i.e., current logic): where you hope the the queue clears up while you're doing your processing.

Reply:
StreamSubscription::enqueueStreamEntry actually waits for a short period for the notification thread to consume the queue. This approach is trying to balance between differnce scenarios and consumption speed.